browser/signin/brave_identity_manager_factory.cc

```diff
@@ -0,0 +1,84 @@
+#include "chrome/browser/image_fetcher/image_decoder_impl.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/signin/account_consistency_mode_manager.h"
+#include "chrome/browser/signin/chrome_signin_client_factory.h"
+#include "components/signin/public/identity_manager/identity_manager_builder.h"
+#include "components/signin/public/webdata/token_web_data.h"
+#include "content/public/browser/network_service_instance.h"
+
+#if !defined(OS_ANDROID)
+#include "chrome/browser/content_settings/cookie_settings_factory.h"
+#include "chrome/browser/web_data_service_factory.h"
+#include "components/content_settings/core/browser/cookie_settings.h"
+#include "components/keyed_service/core/service_access_type.h"
+#include "components/signin/core/browser/cookie_settings_util.h"
+#endif
+
+BraveIdentityManagerFactory::BraveIdentityManagerFactory()
+    : IdentityManagerFactory() {}
+
+BraveIdentityManagerFactory::~BraveIdentityManagerFactory() {}
+
+// static
+signin::BraveIdentityManager* BraveIdentityManagerFactory::GetForProfile(
+    Profile* profile) {
+  return static_cast<signin::BraveIdentityManager*>(
+      GetInstance()->GetServiceForBrowserContext(profile, true));
+}
+
+// static
+signin::BraveIdentityManager*
+BraveIdentityManagerFactory::GetForProfileIfExists(const Profile* profile) {
+  return static_cast<signin::BraveIdentityManager*>(
+      GetInstance()->GetServiceForBrowserContext(const_cast<Profile*>(profile),
+                                                 false));
+}
+
+// static
+BraveIdentityManagerFactory* BraveIdentityManagerFactory::GetInstance() {
+  return base::Singleton<BraveIdentityManagerFactory>::get();
+}
+
+KeyedService* BraveIdentityManagerFactory::BuildServiceInstanceFor(
+    content::BrowserContext* context) const {
+  Profile* profile = Profile::FromBrowserContext(context);
+
+  signin::IdentityManagerBuildParams params;
+  params.account_consistency =
+      AccountConsistencyModeManager::GetMethodForProfile(profile),
+  params.image_decoder = std::make_unique<ImageDecoderImpl>();
+  params.local_state = g_browser_process->local_state();
+  params.network_connection_tracker = content::GetNetworkConnectionTracker();
+  params.pref_service = profile->GetPrefs();
+  params.profile_path = profile->GetPath();
+  params.signin_client = ChromeSigninClientFactory::GetForProfile(profile);
+
+#if !defined(OS_ANDROID)
+  params.delete_signin_cookies_on_exit =
+      signin::SettingsDeleteSigninCookiesOnExit(
+          CookieSettingsFactory::GetForProfile(profile).get());
+  params.token_web_data = WebDataServiceFactory::GetTokenWebDataForProfile(
+      profile, ServiceAccessType::EXPLICIT_ACCESS);
+#endif
+
+  std::unique_ptr<signin::BraveIdentityManager> brave_identity_manager =
+      signin::BuildBraveIdentityManager(&params);
+
+  return brave_identity_manager.release();
+}
+
+void BraveIdentityManagerFactory::RegisterProfilePrefs(
+    user_prefs::PrefRegistrySyncable* registry) {}
```

Comment:
what is the reason that we don't call `signin::IdentityManager::RegisterProfilePrefs`?

Reply:
it crashes with CHECK this pref already registered:
```
[24891:24891:1216/194905.136945:FATAL:pref_registry.cc(65)] Check failed: !defaults_->GetValue(path, nullptr). Trying to register a previously registered pref: google.services.signin_scoped_device_id
#0 0x7f9049d54799 base::debug::CollectStackTrace()
#1 0x7f9049c63813 base::debug::StackTrace::StackTrace()
#2 0x7f9049c82083 logging::LogMessage::~LogMessage()
#3 0x7f9049c829fe logging::LogMessage::~LogMessage()
#4 0x7f904346fad6 PrefRegistry::RegisterPreference()
#5 0x7f90434703ca PrefRegistrySimple::RegisterStringPref()
#6 0x55bfc5931521 ProfileOAuth2TokenService::RegisterProfilePrefs()
#7 0x55bfc5910b5e signin::IdentityManager::RegisterProfilePrefs()
#8 0x7f904607854c DependencyManager::RegisterPrefsForServices()
#9 0x55bfc66d1678 RegisterProfilePrefs()
#10 0x55bfc66cd5f8 ProfileImpl::LoadPrefsForNormalStartup()
#11 0x55bfc66cd234 ProfileImpl::ProfileImpl()
#12 0x55bfc5f1b0ee BraveProfileImpl::BraveProfileImpl()
#13 0x55bfc66cc642 Profile::CreateProfile()
```

Reply:
ah, right. There is already a IdentityManager running